packages/firestore/test/integration/api/query.test.ts

```diff
@@ -1333,6 +1339,626 @@ apiDescribe('Queries', persistence => {
+      const testDocs = {
+        doc1: { key: 'a', sort: 5 },
+        doc2: { key: 'aa', sort: 4 },
+        doc3: { key: 'b', sort: 3 },
+        doc4: { key: 'b', sort: 2 },
+        doc5: { key: 'bb', sort: 1 }
+      };
+      return withTestCollection(persistence, testDocs, async coll => {
+        const snapshot1 = await getDocs(
+          query(
+            coll,
+            where('sort', '>=', 1),
+            where('key', '!=', 'a'),
+            where(documentId(), '<', 'doc5')
+          )
+        );
+        // Document Key in inequality field will implicitly ordered to the last.
+        // Implicitly ordered by: 'key' asc, 'sort' asc, __name__ asc
+        expect(toDataArray(snapshot1)).to.deep.equal([
+          { key: 'aa', sort: 4 },
+          { key: 'b', sort: 2 },
+          { key: 'b', sort: 3 }
+        ]);
+
+        const snapshot2 = await getDocs(
+          query(
+            coll,
+            where(documentId(), '<', 'doc5'),
+            where('sort', '>=', 1),
+            where('key', '!=', 'a')
+          )
+        );
+        // Changing filters order will not effect implicit order.
+        // Implicitly ordered by: 'key' asc, 'sort' asc, __name__ asc
+        expect(toDataArray(snapshot2)).to.deep.equal([
+          { key: 'aa', sort: 4 },
+          { key: 'b', sort: 2 },
+          { key: 'b', sort: 3 }
+        ]);
+
+        const snapshot3 = await getDocs(
+          query(
+            coll,
+            where(documentId(), '<', 'doc5'),
+            where('sort', '>=', 1),
+            where('key', '!=', 'a'),
+            orderBy('sort', 'desc')
+          )
+        );
+        // Ordered by: 'sort' desc,'key' desc,  __name__ desc
+        expect(toDataArray(snapshot3)).to.deep.equal([
+          { key: 'aa', sort: 4 },
+          { key: 'b', sort: 3 },
+          { key: 'b', sort: 2 }
+        ]);
+      });
+    });
+
+    it('can get documents while offline', () => {
+      const testDocs = {
+        doc1: { key: 'a', sort: 1 },
+        doc2: { key: 'aa', sort: 4 },
+        doc3: { key: 'b', sort: 3 },
+        doc4: { key: 'b', sort: 2 }
+      };
+      return withTestCollection(
+        persistence.toLruGc(),
+        testDocs,
+        async (coll, db) => {
+          const query_ = query(
+            coll,
+            where('key', '!=', 'a'),
+            where('sort', '<=', 3)
+          );
+          //populate the cache.
+          const snapshot1 = await getDocs(query_);
+          expect(snapshot1.size).to.equal(2);
+
+          await disableNetwork(db);
+
+          const snapshot2 = await getDocs(query_);
+          expect(snapshot2.metadata.fromCache).to.be.true;
+          expect(snapshot2.metadata.hasPendingWrites).to.be.false;
+          // Implicitly ordered by: 'key' asc, 'sort' asc, __name__ asc
+          expect(toDataArray(snapshot2)).to.deep.equal([
+            { key: 'b', sort: 2 },
+            { key: 'b', sort: 3 }
```

Comment:
unless the data inside the documents are being mutated (which they're not), isn't it easier to do the comparison on the document names, rather than document content? For example, for this one it'd be:

```js
expect(toIds(snapshot2)).to.deep.equal([doc3, doc4]);
```
?

Applies everywhere above.